File: backend/src/main/java/org/d3kad3nt/sunriseClock/backend/data/remote/common/ApiResponse.java
/*
 * (c) Copyright 2025 Palantir Technologies Inc. All rights reserved.
 */

package org.d3kad3nt.sunriseClock.backend.data.remote.common;

import java.io.IOException;
import retrofit2.Response;

// TODO: Make this class more generic / usable for other endpoints (instead of only supporting
// retrofit's Response type). This could be achieved by shifting the logic to decide whether
// a request was successful or not (Type ApiSuccessResponse or ApiErrorResponse) to the specific
// endpoint. Deconz or MQTT endpoint should be able to implement this logic in their request or
// response processing logic!

/**
 * Common class used by API responses.
 *
 * <p>Adapted from the official Google architecture-components github-sample app under
 * https://github.com/android/architecture-components-samples/blob/master/GithubBrowserSample/app/src/main/java/com
 * /android/example/github/api/ApiResponse.kt.
 */
public abstract class ApiResponse<T> {

    public static <T> ApiResponse<T> create(Throwable error) {
        if (error.getMessage() != null) {
            return new ApiErrorResponse<>(error.getMessage());
        } else {
            return new ApiErrorResponse<>("Unknown error!");
        }
    }

    public static <T> ApiResponse<T> create(final Response<T> response) {
        if (response.isSuccessful()) {
            final T body = response.body();
            if (body == null || response.code() == 204) {
                return new ApiEmptyResponse<>();
            } else {
                return new ApiSuccessResponse<>(body);
            }
        } else {
            String msg = null;
            if (response.errorBody() != null) {
                try {
                    msg = response.errorBody().string();
                } catch (IOException ignored) {
                }
            }
            final String errorMsg;
            if ((msg == null) || msg.isEmpty()) {
                if (response.message().isEmpty()) {
                    errorMsg = "Unknown error!";
                } else {
                    errorMsg = response.message();
                }
            } else {
                errorMsg = msg;
            }
            return new ApiErrorResponse<>(errorMsg);
        }
    }
}

File: app/src/main/java/org/d3kad3nt/sunriseClock/data/remote/common/ApiResponse.kt
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.d3kad3nt.sunriseClock.data.remote.common

import retrofit2.Response

//TODO: Make this class more generic / usable for other endpoints (instead of only supporting
// retrofit's Response type). This could be achieved by shifting the logic to decide whether
// a request was successful or not (Type ApiSuccessResponse or ApiErrorResponse) to the specific
// endpoint. Deconz or MQTT endpoint should be able to implement this logic in their request or
// response processing logic!

/**
 * Common class used by API responses.
 * Copied from the official Google architecture-components github-sample under https://github.com/android/architecture-components-samples/blob/master/GithubBrowserSample/app/src/main/java/com/android/example/github/api/ApiResponse.kt
 *
 * @param <T> the type of the response object
 */
@Suppress("unused") // T is used in extending classes
sealed class ApiResponse<T> {
    companion object {
        fun <T> create(error: Throwable): ApiErrorResponse<T> {
            return ApiErrorResponse(error.message ?: "unknown error")
        }

        fun <T> create(response: Response<T>): ApiResponse<T> {
            return if (response.isSuccessful) {
                val body :T? = response.body()
                if (body == null || response.code() == 204) {
                    ApiEmptyResponse()
                } else {
                    ApiSuccessResponse(
                            body = body
                    )
                }
            } else {
                val msg: String?  = response.errorBody()?.string()
                val errorMsg: String? = if (msg.isNullOrEmpty()) {
                    response.message()
                } else {
                    msg
                }
                ApiErrorResponse(errorMsg ?: "unknown error")
            }
        }
    }
}

/**
 * Separate class for HTTP 204 responses so that we can make ApiSuccessResponse's body non-null.
 */
class ApiEmptyResponse<T> : ApiResponse<T>()

data class ApiSuccessResponse<T>(val body: T) : ApiResponse<T>()

data class ApiErrorResponse<T>(val errorMessage: String) : ApiResponse<T>()
